### ISSUE TITLE
Big Picture Mode loses focus after exiting a game

### ISSUE THREAD

**Issue by Enigma0** [ValveSoftware/steam-for-linux#4611]
#### Your system information

* Steam client version: Latest (non beta)
* Distribution (e.g. Ubuntu): Ubuntu GNOME 16.04.1
* Opted into Steam client beta?: [No]
* Have you checked for system updates?: [Yes]

#### Please describe your issue in as much detail as possible:
Re-opening this issue: https://github.com/ValveSoftware/steam-for-linux/issues/2050 by making a new issue since that one was seemingly fixed.

#### Steps for reproducing this issue:

1. Launch Big Picture Mode
2. Launch a game
3. Exit game
4. Big Picture Mode does not regain focus

**Comment by mattbell87** [comment]
Happens to me on Fedora 24 and did the same when I was on 23, both on stock Gnome3. I have to Alt+Tab every time I quit a game.

A side note, I also often get the screen appearing off centre after I quit a game. I can see my Gnome wallpaper behind Big Picture after I switch back. Seems to happen more if I use a lower resolution in Steam than my native. I'm not sure if that is related to this bug or if it belongs in another issue.

**Comment by kerbastrar** [comment]
Same issue for me with Arch Linux and KDE 5.7

Also noticing that the standard Steam desktop notification and Big Picture notification show up at the same time while playing a game. Sounds like something is going on with the main Steam window while the game/stream window is in the foreground

**Comment by parkerlreed** [comment]
Same on Arch Linux with Gnome and the latest Steam beta 1481318740

**Comment by mattbell87** [comment]
Still happening with Fedora which uses Gnome by default.

Kind of annoying for me as I don't have a keyboard or mouse as it's just a
Mac Mini plugged in to my TV. I always use "XMouse" on my Android to
Alt+Tab back to steam.

**Comment by HoratiuMl** [comment]
This issue was present on my Arch with GNOME for more than a year.

I have a Steam Controller and the mouse functionality helps but not much, because sometimes the dash (GNOME's taskbar) won't show up until I alt tab. The focus is obviously not set on Steam but I don't think it's set on the desktop either, I don't know what happens. It only works if I alt tab to something else (e.g. the file explorer if one is open, or Chrome) and then everything is usable and I can alt-tab back to Steam.

Even more annoying is that sometimes after I do that (alt-tabbing back to Steam) I can see a gap where the GNOME Panel would have been (on top) and I can see my desktop wallpeper through it and Steam only starts displaying below it (even tho it's still in "fullscreen") thus resulting in the bottom part being outside of the display area and not visible.

**Comment by parkerlreed** [comment]
I'm not sure when or what update but Big Picture is correctly coming into focus after exiting games now. using both the exit menu in the overlay and exiting a game from its own menu.

**Comment by danielbibit** [comment]
Same problema for me on Fedora 25 with gnome 3. Using a Wayland sessions, I can't even alter tab to steam back without struggling, and the steam window do not goes full screen,  the windows appears on the bottom right side of the screen. In xorg session, same bug, but alt tab goes back to big picture full screen.

**Comment by Cybolic** [comment]
Same issue for me on Ubuntu Gnome. It makes my Steam Link extremely cumbersome as after each game I have to go into the other room and switch windows on the computer before I can continue.

**Comment by parkerlreed** [comment]
@Cybolic Any controller now should operate the mouse with the right stick/touchpad. Does that not work to switch back to it?

**Comment by Cybolic** [comment]
@parkerlreed No, the Steam Controller goes back to the standard Big Picture game pad setup, even when used as a mouse in-game.

**Comment by parkerlreed** [comment]
@Cybolic Ok weird. For me if Big Picture fails to regain focus, the Steam Controller dumps back to the desktop mapping and lets me navigate that way.

**Comment by shaohme** [comment]
I run steam on Debian Stretch with I3 window manager.
When running in big-picture mode it starts in fullscreen as it should, but after exiting from a game, steam becomes a window/tab and I need to manually "fullscreen" it again with Win+F key.

**Comment by pnegry** [comment]
I solved this using the following script launched from Kodi as a pre-command to Steam

```
sleep 10
c=0
while pgrep /usr/bin/steam > /dev/null
do
  DISPLAY=:0 wmctrl -l |grep Stream
  if [ $? == 0 ]
  then
    echo Switching focus to stream window
    DISPLAY=:0 wmctrl -a Stream
  elif [ `DISPLAY=:0 wmctrl -l |grep -i kodi` ]
  then
   echo Kodi has focus
   break
  else
    DISPLAY=:0 wmctrl -l |grep  steam
    if [ $? == 1 ]
    then
      echo Waiting up to 30 seconds for steam window to appear
      if [ $c == 3 ]
      then
        killall steam
      else
        c=`expr $c + 1`
        sleep 5
      fi
    else
      echo Switching focus to Steam BPM
      DISPLAY=:0 wmctrl -a Steam
    fi
  fi
  sleep 5
done
```

**Comment by oblitum** [comment]
Happens for me as well, it destroys the couch experience.

**Comment by oblitum** [comment]
This is surely an issue with the notification system. I've managed to disable the `"Steam" is ready` notification on Gnome 3 by disabling all notifications in the control panel and restarting the shell (ALT+F2). When I exit a steam game, the notification didn't appear, and the focus was not lost, but... exiting it a second time the notification still popped up, and the client focus was lost again.

**Comment by oblitum** [comment]
I was unable to block this notification even by using a [do-not-disturb-button](https://extensions.gnome.org/extension/964/do-not-disturb-button/) extension.

**Comment by Enigma0** [comment]
Such a pointless notification also... Is Steam creating it?

**Comment by oblitum** [comment]
@Enigma0 can't say, I'm about to try other extensions https://extensions.gnome.org/extension/1007/window-is-ready-notification-remover/.

**Comment by oblitum** [comment]
@Enigma0 Indeed, https://github.com/nunofarruca/WindowIsReady_Remover has fixed it :)

**Comment by oblitum** [comment]
@Enigma0 Solved "mostly", there are specific circumstances that does not bring the focus back, related with the way I exit the game... can't describe exactly what though. I've noticed that exiting from Guide controller button is acting correctly, but for some in game menu exits may or may not work.

**Comment by oblitum** [comment]
@Enigma0 OK, this one works better than the last, fixed it in any case for me :) https://github.com/tak0kada/gnome-shell-extension-stealmyfocus

**Comment by Enigma0** [comment]
Yet another essential extension vs having baked in functionality in GNOME... If it works, that's great though.

**Comment by redpop1** [comment]
Using Ubuntu 16.04 without gnome-extensions but unity instead.

Same problem, steam loses focus after exiting a game.

Is there a way around without gnome-extensions?

**Comment by kisak-valve** [comment]
Hello @redpop1, running steam with `SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0` set should also workaround this issue (from #4769).

**Comment by Mattia98** [comment]
How is this still not fixed yet? Still happens to me on Ubuntu MATE 17.10 and 16.04 and Xubuntu 17.04
C'mon guys... I'm gonna try kisak's workaround soon.

**Comment by shooter0106** [comment]
Same problem on Debian Stretch with Gnome desktop.

**Comment by redpop1** [comment]
Sorry for the late reply: Kisak solution is indeed a workaround. Thanks Kisak.

**Comment by lkmcdo** [comment]
Are people running into this issue with KDE as well?

**Comment by oblitum** [comment]
I can confirm @kisak-valve's solution also works for me on Gnome3/Archlinux, no need for focus extensions anymore, finally.

**Comment by Mattia98** [comment]
It seems to be fixed now for me. Works out of the box on Xubuntu 18.04 no workarounds.

**Comment by oblitum** [comment]
At least on lastest Arch/Gnome it's not, but the flag fixes it.

**Comment by oblitum** [comment]
Ops, it seems there are some drawbacks about this setting (`SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0`):

1. **On i3**, it happened that when most games are started, their screen doesn't come up, they stay behind the BigPicture GUI, I have to move it to another workspace to see the game window.
2. It looks like BigPicture stays running consuming CPU.

**Comment by lucarillosi** [comment]
Same problem on KDE for me... Any updates guys?

**Comment by shazzner** [comment]
I have the problem as well, especially annoying as it breaks Steam Link. When I exit from the couch I just see a blank desktop.

I'll try the fix outlined above.

**Comment by FrankieEnc** [comment]
Yeah, I have this too. Annoying.

**Comment by yaomtc** [comment]
The environment variable works for me both on Ubuntu (Gnome) and Solus (Budgie). No extension needed here. I have `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U` after every `[Exec]` in `/usr/share/applications/steam.desktop`. It may make sense to update the default .desktop file that Steam provides, unless t

**Comment by afettouhi** [comment]
Small update on this via the similar KDE bug that I originally reported. See this specifically this comment

https://bugs.kde.org/show_bug.cgi?id=359327#c28

**Comment by afettouhi** [comment]
Seems the conclusion in the kde.org bug report is that Steam seems to change _NET_WM_STATE for some reason.

**Comment by Mohandevir** [comment]
My workaround is to change de default desktop behaviour of the Steam Controller mapping to change the ALT button to ALT+F4. It brings back Steam BPM. Maybe something similar may be done for other controllers or scripted (?) at game exit?

Hope it helps... Sorry if not. :)

**Comment by NerosTie** [comment]
I can confirm it works with `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U`. You don't need to modify the .desktop file, just replace the command in your applications manager.

On KDE: right click on start menu then "edit applications" then Steam/Steam(Native) and replace the command with:

`env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U` for Steam
`env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam-native %U` for Steam(Native)

**Comment by afettouhi** [comment]
> I can confirm it works with `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U`. You don't need to modify the .desktop file, just replace the command in your applications manager.
> 
> On KDE: right click on start menu then "edit applications" then Steam/Steam(Native) and replace the command with:
> 
> `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U` for Steam
> `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam-native %U` for Steam(Native)
> 
> And click on Save.

Can confirm that this workaround works on Arch Linux with KDE.

**Comment by guustflater** [comment]
Works for me! What I don't understand is that a bug from 2016 which is so easy to fix in the next steam update is still here. Please fix this for other users Valve! 

Otherwise, thank you very much for the solution and keep up the good work! :)

**Comment by noobmagnet** [comment]
Still happens for me with fresh install of Ubuntu 19.10, Steam /w beta opted in, all updated and checked, and when I quit out of Dark Souls, or Doom 2016, it stays minimized from big picture mode. (10/20/2019)

**Comment by Enigma0** [comment]
Same here. Happens consistently - very annoying.

**Comment by noobmagnet** [comment]
> > I can confirm it works with `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U`. You don't need to modify the .desktop file, just replace the command in your applications manager.
> > On KDE: right click on start menu then "edit applications" then Steam/Steam(Native) and replace the command with:
> > `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam %U` for Steam
> > `env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/bin/steam-native %U` for Steam(Native)
> > And click on Save.
> 
> Can confirm that this workaround works on Arch Linux with KDE.

This works for me thanks, except, at least in Ubuntu 19,10, the path to steam is `/usr/games/steam %U` instead.

**Comment by TTimo** [comment]
CNR on Manjaro + KDE (Plasma). I will check on an Ubuntu.

**Comment by vasiliy0s** [comment]
https://extensions.gnome.org/extension/1005/focus-my-window/ - works for me on Ubuntu 18.04

**Comment by Neverest** [comment]
Still happening here on Ubuntu Mate 18.04 & 20.04 versions.

**Comment by thewoza** [comment]
env SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS=0 /usr/games/steam %U works with ElementaryOS (Ubuntu 18.04 base). It's also had the added bonus of focusing all games on launch which I was having a few problems with too.

**Comment by reedlove** [comment]
Hi there.
We ever going to get a fix, not a workaround, for this four year old bug @Plagman @gdrewb-valve ?
I know that Linux can be a fractured pain-in-the-ass but there has to be some universal mechanism for regaining focus under X11 at least, regardless of desktop manager or compositor...

**Comment by DanMan** [comment]
I don't have this problem anymore on a freshly installed Fedora 32 with Gnome 3.36.4.

**Comment by reedlove** [comment]
It still has issues in Ubuntu 18.04, Ubuntu 20.04, Pop!_OS 20.04, and elementary OS 5.1.6 (based on Ubuntu 18.04 HWE) using either an nVidia GTX 1060 or GT 1030. =/

**Comment by troywweber7** [comment]
Yup, this is happening to me, too.

**Comment by farukr1** [comment]
Fedora 35 and still the same issue.

**Comment by DoctorWhoof** [comment]
It is almost 2023, and I still have this issue on some games running Big Picture mode on a windows 11 machine that sits in the basement, and is controlled solely with an Xbox controller.

It's a real problem since I can't regain control of Steam using the controller. I have to remote login (via Moonlight), move the mouse a bit, and only then I can use the controller again.

Weirdly enough, I can hear the UI sounds while using the controller, but the screen is a solid blue and I can't see what's happening until the mouse is moved (using the remote viewer trick).

This may be helpful though: If I quit the game by clicking the big round "X" button on the middle of the controller to invoke the Steam menu overlay, and then quit using Steam's  "Exit Game" option it works fine. It's when the game quits "by itself" that Steam seems unaware that the game has finished running.

Thanks.

**Comment by DoctorWhoof** [comment]
I'm on Windows, not Linux , but I still have this issue on some games running Big Picture mode on a windows 11 machine that sits in the basement, and is controlled solely with an Xbox controller.

It's a real problem since I can't regain control of Steam using the controller. I have to remote login (via Moonlight), move the mouse a bit, and only then I can use the controller again.

Weirdly enough, I can hear the UI sounds while using the controller, but the screen is a solid blue and I can't see what's happening until the mouse is moved (using the remote viewer trick).

This may be helpful though: If I quit the game by clicking the big round "X" button on the middle of the controller to invoke the Steam menu overlay, and then quit using Steam's  "Exit Game" option it works fine. It's when the game quits "by itself" that Steam seems unaware that the game has finished running.

Thanks.

**Comment by reedlove** [comment]
Still happens here too. Gnome on Arch. Nvidia proprietary graphics.

**Comment by ns6000** [comment]
> I'm on Windows, not Linux , but I still have this issue on some games running Big Picture mode on a windows 11 machine that sits in the basement and has an HDMI cable to my living room TV, and is controlled solely with an Xbox controller.
> 
> It's a real problem since I can't regain control of Steam using the controller. I have to remote login (via Moonlight), move the mouse a bit, and only then I can use the controller again.
> 
> Weirdly enough, I can hear the UI sounds while using the controller, but the screen is a solid blue and I can't see what's happening until the mouse is moved (using the remote viewer trick).
> 
> This may be helpful though: If I quit the game by clicking the big round "X" button on the middle of the controller to invoke the Steam menu overlay, and then quit using Steam's "Exit Game" option it works fine. It's when the game quits "by itself" that Steam seems unaware that the game has finished running.
> 
> Thanks.

(sorry on commenting in steam-for-linux repo but steam for windows has none)
Exactly my experience, also on Windows 11, new Big Picture is in majority of times unable to distinguish that the game has exited unless the game is killed from within the new Big Picture itself...

**Comment by jhpgcj** [comment]
Confirming this also happens to me on Pop!_OS 22.04. However, sometimes, Big Picture get the focus right. So it's kind of a hit-or-miss problem.
